### PR TITLE
#2089 - Use tolerance in membership test of Polyhedron

### LINKED ISSUE #2089: Use tolerance in membership test of Polyhedron
https://github.com/JuliaReach/LazySets.jl/blob/894f77bd2309a80e02d5be50acfb6dbc7c617947/src/Interfaces/AbstractPolyhedron_functions.jl#L50

### PR DESCRIPTION
**Pull request by SebastianGuadalupe**
Closes #2089.

### CODE REVIEW

**Comment by mforets** [comment]
```julia
LazySets.Polyhedron: Test Failed at /home/travis/build/JuliaReach/LazySets.jl/test/unit_Polyhedron.jl:136
  Expression: an_element(P) ∈ P
   Evaluated: [Inf, Inf] ∈ HPolyhedron{Float64,Array{Float64,1}}(HalfSpace{Float64,Array{Float64,1}}[HalfSpace{Float64,Array{Float64,1}}([0.06, -0.4], 1.0), HalfSpace{Float64,Array{Float64,1}}([-0.02, 0.1], -1.0)])
Stacktrace:
 [1] top-level scope at /home/travis/build/JuliaReach/LazySets.jl/test/unit_Polyhedron.jl:136
 [2] include(::String) at ./client.jl:439
 [3] macro expansion at /home/travis/build/JuliaReach/LazySets.jl/test/runtests.jl:85 [inlined]
 [4] macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113 [inlined]
 [5] macro expansion at /home/travis/build/JuliaReach/LazySets.jl/test/runtests.jl:85 [inlined]
 [6] macro expansion at ./util.jl:175 [inlined]
 [7] top-level scope at /home/travis/build/JuliaReach/LazySets.jl/test/runtests.jl:85
```

**Comment by schillic** [comment]
This is the failing test:
```julia
julia> P = HPolyhedron([HalfSpace(N[3//50, -4//10], N(1)),
                        HalfSpace(N[-1//50, 1//10], N(-1))]);

julia> an_element(P)
2-element Array{Float64,1}:
 Inf
 Inf

julia> an_element(P) ∈ P
false
```

**Comment by schillic** [comment]
Checking for NaN imitates the previous behavior. I am not convinced that this behavior was good but we can address this later.

**Comment by SebastianGuadalupe** [comment]
> Checking for NaN imitates the previous behavior. I am not convinced that this behavior was good but we can address this later.

Are there other cases where the dot product gives you `NaN` other than when the point checked is `Inf`?

**Comment by schillic** [comment]
The problem is that in the example below I would argue that the point should not be contained in `P1`.

```julia
julia> P1 = HPolyhedron([HalfSpace([1.0, 0.0], 0.0)]);

julia> P2 = HPolyhedron([HalfSpace([-1.0, 0.0], 0.0)]);

julia> [Inf, Inf] ∈ P1
true

julia> [Inf, Inf] ∈ P2
true
```

Note that in 1D the behavior is as I would expect it:

```julia
julia> P1 = HPolyhedron([HalfSpace([1.0], 0.0)]);

julia> P2 = HPolyhedron([HalfSpace([-1.0], 0.0)]);

julia> [Inf] ∈ P1
false

julia> [Inf] ∈ P2
true
```

But as I said this is the behavior in `master`, so we can merge this.

**Comment by mforets** [comment]
Can we mark the test as broken?  This can be done with `@test_broken`.

The transformation rule/convention states that

`dot(c.a, x) > c.b <=> !_leq(dp, c.b)`.

**Comment by schillic** [comment]
We can just use a vector with finite entries.

**Comment by mforets** [comment]
> The problem is that in the example below I would argue that the point should not be contained in P1.

yes, i agree. this is in fact an edge case because

```julia
P1 = HPolyhedron([HalfSpace([1.0, 0.0], 0.0)]) # x <= 0
dot(P1.constraints[1].a, [Inf, Inf])
NaN

[Inf, Inf] ∈ P1
true
```

however note that

```julia
P1 = HPolyhedron([HalfSpace([1.0, 1e-20], 0.0)]) # x <= 0
dot(P1.constraints[1].a, [Inf, Inf])
Inf

[Inf, Inf] ∈ P1
false
```

**Comment by mforets** [comment]
in sum i think that making such runtime checks to handle special cases like 0 * Inf is not a good idea (at least, for the "default" behavior of this function).

so i propose to:

- rervert commit "check for NaN"
- keep the test with infinities but mark it as currently consistently broken, `@test_broken !([Inf, Inf] ∈ P1)`
- add a new test that passes, with finite entries 
- add a new test that passes, with infinite entries but non-zero normal vector

**Comment by mforets** [comment]
@schillic  does that list make sense to you?

**Comment by schillic** [comment]
I would be fine with items 1 and 3 only but as you prefer.

**Comment by schillic** [comment]
Still missing:
> * add a new test that passes, with finite entries

**Comment by SebastianGuadalupe** [comment]
There is already a test with finite entries:

**Comment by mforets** [comment]
ok, thanks for checking. then this lgtm after tests pass

File: test/unit_Polyhedron.jl [hunk 1]
```diff
@@ -130,10 +130,13 @@ for N in [Float64]
...
     @test σ(d, p) == N[0, 0]
 
+    # membership
+    @test [Inf, Inf] ∉ p
+
     # an_element
     P = HPolyhedron([HalfSpace(N[3//50, -4//10], N(1)),
                      HalfSpace(N[-1//50, 1//10], N(-1))])
-    @test an_element(P) ∈ P
+    @test_broken an_element(P) ∈ P
```
Comment (mforets) [inline test/unit_Polyhedron.jl:139]:
```suggestion
    @test_broken an_element(P) ∈ P # see LazySets.jl/pull/2197
```

**Comment by mforets** [comment]
tests passed. thanks @SebastianGuadalupe !
